src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -864,7 +864,36 @@ internal static ImmutableArray<T> AddRange<T>(this ImmutableArray<T> self, in Te
             return builder.ToImmutableAndFree();
         }
 
-        internal static bool HasDuplicates<T>(this ImmutableArray<T> array, IEqualityComparer<T>? comparer = null)
+        internal static bool HasDuplicates<T>(this ImmutableArray<T> array)
+        {
+            switch (array.Length)
+            {
+                case 0:
+                case 1:
+                    return false;
+
+                case 2:
+                    return EqualityComparer<T>.Default.Equals(array[0], array[1]);
+
+                default:
+                    var set = PooledHashSet<T>.GetInstance();
+                    var foundDuplicate = false;
+
+                    foreach (var element in array)
+                    {
+                        if (!set.Add(element))
+                        {
+                            foundDuplicate = true;
+                            break;
+                        }
+                    }
+
+                    set.Free();
+                    return foundDuplicate;
+            }
+        }
+
+        internal static bool HasDuplicates<T>(this ImmutableArray<T> array, IEqualityComparer<T> comparer)
```

Comment:
>HasDuplicates [](http://example.com/codeflow?start=29&length=13)

The difference in usage of the poll between the two overloads bothers me somewhat. Consider documenting the fact in a doc comment, recommending to use the other overload when the comparer used is the default one.